src/services/item/plugins/file/service.ts

```diff
@@ -130,20 +122,17 @@ class FileItemService {
 
       // create item from file properties
       const name = filename.substring(0, ORIGINAL_FILENAME_TRUNCATE_LIMIT);
+      const fileProperties: FileItemProperties = { name: filename, path: filepath, mimetype, size };
       const item = {
         name,
         description,
         type: this.fileService.type,
         extra: {
-          [this.fileService.type]: {
-            name: filename,
-            path: filepath,
-            mimetype,
-            size,
-          },
-          // todo: fix type
-        } as any,
-        parentId,
+          // this is needed because if we directly use `this.fileService.type` then TS widens the type to `string` which we do not want
+          ...(this.fileService.type === ItemType.LOCAL_FILE
+            ? { [ItemType.LOCAL_FILE]: fileProperties }
+            : { [ItemType.S3_FILE]: fileProperties }),
+        } as ItemExtraMap[typeof this.fileService.type],
```

Comment:
Can't we restrict the type of  `fileService.type` instead?

Reply:
This one was a tricky one too. You can't use `fileService.type` as the key here because it the union of literal `"s3File" | "localFile"`. So it will not result in the expected type of FileItemProperties ... The only way to make TS abid to the fact that it was indeed one or the other was by explicitly checking for the value and setting the value again explicitly like I did here with the literal.